lib/utils/agentconn/agent_windows.go

```diff
@@ -20,20 +20,240 @@ limitations under the License.
+		return nil, trace.Wrap(err)
+	}
+
+	var uid int64
+	var unsureOfUID bool
+	if !strings.HasPrefix(u.Uid, wellKnownSIDPrefix) {
+		// the format of the SID isn't supported, fallback to getting
+		// the UID from 'ps'
+		uid, err = getCygwinUIDFromPS()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else {
+		// Attempt to get a Cygwin UID from a Windows SID. Details of
+		// UID -> SID mapping here: https://cygwin.com/cygwin-ug-net/ntsec.html
+		sidParts := strings.Split(u.Uid, "-")
+		if len(sidParts) == 4 {
+			// well-known SIDs in the NT_AUTHORITY domain of the S-1-5-RID type
+			uid, err = strconv.ParseInt(sidParts[3], 10, 32)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		} else if len(sidParts) == 5 {
+			// other well-known SIDs that aren't groups
+			x, err := strconv.ParseInt(sidParts[3], 10, 32)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			rid, err := strconv.ParseInt(sidParts[4], 10, 32)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			uid = 0x1000*x + rid
+		} else if len(sidParts) == 8 {
+			// SIDs from the local machine's SAM, the machine's primary
+			// domain, or a trusted domain of the machine's primary domain
+			uid, err = strconv.ParseInt(sidParts[7], 10, 32)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			unsureOfUID = true
+		} else {
+			// the format of the SID isn't supported, fallback to getting
+			// the UID from 'ps'
+			uid, err = getCygwinUIDFromPS()
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+	}
+
+	// dial socket and complete handshake
+	var conn net.Conn
+	if !unsureOfUID {
+		// we're confident in what the Cygwin UID is, only make one attempt
+		// at establishing a connection
+		conn, err = attemptCygwinHandshake(port, key, uid)
+		if err == nil {
+			return conn, nil
+		}
+	} else {
+		// the Cygwin UID could be built a few different ways; attempt
+		// with all UIDs until one succeeds
+		cygwinRIDNums := []int64{0x30000, 0x100000, 0x80000000}
+		for _, num := range cygwinRIDNums {
+			conn, err = attemptCygwinHandshake(port, key, num+uid)
+			if err == nil {
+				return conn, nil
+			}
+		}
+
+		// none of those UIDs worked, fallback to getting UID from 'ps'
+		uid, err = getCygwinUIDFromPS()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		conn, err = attemptCygwinHandshake(port, key, uid)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return conn, nil
+}
+
+// use Cygwin 'ps' binary to get the Cygwin UID of the current user
+func getCygwinUIDFromPS() (int64, error) {
+	// Cygwin 'bash' is used to call 'ps' so a Cygwin environment can be
+	// inherited by 'ps'
+	psOutput, err := exec.Command("bash.exe", "-c", "ps").Output()
+	if err != nil {
+		return 0, trace.Wrap(err)
+	}
+	psMatches := psLine.FindStringSubmatch(string(psOutput))
+	if len(psMatches) != 2 {
+		return 0, trace.Errorf("error reading Cygwin ps output")
+	}
+	uid, err := strconv.ParseInt(psMatches[1], 10, 32)
```

Comment:
Why are we parsing as `32` bits if this function returns `int64`?

Reply:
I was only returning an `int64` because that's what `strconv.ParseInt` returned, but I switched to use `strconv.ParseUint` and return a `uint32` instead.